hashing/src/lib.rs

```diff
@@ -1,174 +1,165 @@
+        Digest(result)
+    }
+
+    /// Hashes a pair of byte slices into a single [`Digest`]
+    pub fn hash_pair<T: AsRef<[u8]>, U: AsRef<[u8]>>(data1: T, data2: U) -> Digest {
+        let mut result = [0; Digest::LENGTH];
+        let mut hasher = VarBlake2b::new(Digest::LENGTH).unwrap();
+        hasher.update(data1);
+        hasher.update(data2);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        Digest(result)
+    }
+
+    /// Returns a reference to the underlying value.
+    pub fn inner(&self) -> &[u8; Digest::LENGTH] {
+        &self.0
+    }
+
+    /// Returns the underlying value.
+    pub fn value(&self) -> [u8; Digest::LENGTH] {
         self.0
     }
 
-    /// Returns a copy of the wrapped `u8` array as a `Vec`
+    /// Converts the underlying hash digest array to a `Vec`
+    #[cfg(feature = "std")]
     pub fn into_vec(self) -> Vec<u8> {
         self.0.to_vec()
     }
 
     /// Returns a `Digest` parsed from a hex-encoded `Digest`.
-    pub fn from_hex<T: AsRef<[u8]>>(hex_input: T) -> Result<Self, Error> {
+    #[cfg(feature = "std")]
+    pub fn from_hex<T: AsRef<[u8]>>(hex_input: T) -> Result<Self, hex::FromHexError> {
         let mut inner = [0; Digest::LENGTH];
         hex::decode_to_slice(hex_input, &mut inner)?;
         Ok(Digest(inner))
     }
-
-    /// Generates a random instance using a `TestRng`.
-    #[cfg(test)]
-    pub fn random(rng: &mut TestRng) -> Self {
-        Digest(rng.gen::<[u8; Digest::LENGTH]>())
-    }
 }
 
-impl AsRef<[u8]> for Digest {
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
-    }
-}
-
-impl From<[u8; Digest::LENGTH]> for Digest {
-    fn from(inner: [u8; Digest::LENGTH]) -> Self {
-        Digest(inner)
+#[cfg(feature = "std")]
+impl std::fmt::LowerHex for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let hex_string = base16::encode_lower(self.inner());
+        if f.alternate() {
+            write!(f, "0x{}", hex_string)
+        } else {
+            write!(f, "{}", hex_string)
+        }
     }
 }
 
-impl TryFrom<&[u8]> for Digest {
-    type Error = TryFromSliceError;
-
-    fn try_from(slice: &[u8]) -> Result<Digest, Self::Error> {
-        <[u8; Digest::LENGTH]>::try_from(slice).map(Digest)
+#[cfg(feature = "std")]
+impl std::fmt::UpperHex for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let hex_string = base16::encode_upper(self.inner());
+        if f.alternate() {
+            write!(f, "0x{}", hex_string)
+        } else {
+            write!(f, "{}", hex_string)
+        }
     }
 }
 
-impl Debug for Digest {
-    fn fmt(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
-        write!(formatter, "{}", HexFmt(&self.0))
+#[cfg(feature = "std")]
+impl std::fmt::Display for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:10}", base16::encode_lower(&self.0))
```

Comment:
…and here we wouldn't even have to encode all the bytes, since we're printing only five of them.